File: web/src/metrics.ts
import { Session } from "./session";
import { pack, toString } from "./utils";

const metricsEndpoint = 0x05;

export enum MetricKind {
  counter,
  gauge,
}

export enum MetricType {
  long,
  float,
  double,
}

export interface MetricInfo {
  ref: number;
  kind: MetricKind;
  type: MetricType;
  name: string;
  size: number;
}

export interface MetricLayout {
  keys: string[];
  values: string[][];
}

export async function listMetrics(
  s: Session,
  timeout: number = 5000
): Promise<MetricInfo[]> {
  // send command
  const cmd = pack("o", 0);
  await s.send(metricsEndpoint, cmd, 0);

  // prepare list
  const list: MetricInfo[] = [];

  for (;;) {
    // receive reply or return list on ack
    const [reply, ack] = await s.receive(metricsEndpoint, true, timeout);
    if (ack) {
      break;
    }

    // verify reply
    if (reply.length < 4) {
      throw new Error("Invalid reply");
    }

    // parse reply
    const ref = reply[0];
    const kind = reply[1];
    const type = reply[2];
    const size = reply[3];
    const name = toString(reply.slice(4));

    // append info
    list.push({ ref, kind, type, name, size });
  }

  return list;
}

export async function describeMetric(
  s: Session,
  ref: number,
  timeout: number = 5000
): Promise<MetricLayout> {
  // send command
  const cmd = pack("oo", 1, ref);
  await s.send(metricsEndpoint, cmd, 0);

  // prepare lists
  let keys: string[] = [];
  let values: string[][] = [];

  for (;;) {
    // receive reply
    const [reply, ack] = await s.receive(metricsEndpoint, true, timeout);
    if (ack) {
      break;
    }

    // verify reply
    if (reply.length < 1) {
      throw new Error("Invalid reply");
    }

    // handle key
    if (reply[0] === 0) {
      // verify reply
      if (reply.length < 3) {
        throw new Error("Invalid reply");
      }

      // parse reply
      const num = reply[1];
      const key = toString(reply.slice(2));

      // add key
      keys[num] = key;
      values[num] = [];

      continue;
    }

    // handle value
    if (reply[0] === 1) {
      // verify reply
      if (reply.length < 4) {
        throw new Error("Invalid reply");
      }

      // parse reply
      const numKey = reply[1];
      const numValue = reply[2];
      const value = toString(reply.slice(3));

      // add value
      values[numKey][numValue] = value;

      continue;
    }

    throw new Error("Invalid reply");
  }

  return {
    keys: keys,
    values: values,
  };
}

export async function readMetrics(
  s: Session,
  ref: number,
  timeout: number = 5000
): Promise<Uint8Array> {
  // send command
  const cmd = pack("oo", 2, ref);
  await s.send(metricsEndpoint, cmd, 0);

  // receive reply
  const [reply] = await s.receive(metricsEndpoint, false, timeout);

  return reply;
}

export async function readLongMetrics(
  s: Session,
  ref: number,
  timeout: number = 5000
): Promise<number[]> {
  // receive value
  const reply = await readMetrics(s, ref, timeout);

  // convert reply
  let list: number[] = [];
  for (let i = 0; i < reply.length; i += 4) {
    list.push(new DataView(reply.buffer).getInt32(i, true));
  }

  return list;
}

export async function readFloatMetrics(
  s: Session,
  ref: number,
  timeout: number = 5000
): Promise<number[]> {
  // receive value
  const reply = await readMetrics(s, ref, timeout);

  // convert reply
  let list: number[] = [];
  for (let i = 0; i < reply.length; i += 4) {
    list.push(new DataView(reply.buffer).getFloat32(i, true));
  }

  return list;
}

export async function readDoubleMetrics(
  s: Session,
  ref: number,
  timeout: number = 5000
): Promise<number[]> {
  // receive value
  const reply = await readMetrics(s, ref, timeout);

  // convert reply
  let list: number[] = [];
  for (let i = 0; i < reply.length; i += 8) {
    list.push(new DataView(reply.buffer).getFloat64(i, true));
  }

  return list;
}

File: python/naos/metrics.py
from __future__ import annotations

import struct
from dataclasses import dataclass
from enum import IntEnum
from typing import List

from .session import Session
from .utils import pack

_metrics_endpoint = 0x05


class MetricKind(IntEnum):
    COUNTER = 0
    GAUGE = 1


class MetricType(IntEnum):
    LONG = 0
    FLOAT = 1
    DOUBLE = 2


@dataclass
class MetricInfo:
    ref: int
    kind: MetricKind
    type: MetricType
    name: str
    size: int


@dataclass
class MetricLayout:
    keys: List[str]
    values: List[List[str]]


async def list_metrics(s: Session, timeout: float = 5.0) -> List[MetricInfo]:
    """Return a list of all metrics."""

    # send command
    await s.send(_metrics_endpoint, pack("o", 0), 0)

    # prepare list
    result = []

    while True:
        # receive reply or return list on ack
        reply, ack = await s.receive(_metrics_endpoint, True, timeout)
        if ack:
            break

        # verify reply
        if len(reply) < 4:
            raise RuntimeError("invalid reply")

        # parse reply
        ref, kind, type_, size = reply[0], reply[1], reply[2], reply[3]
        name = reply[4:].decode()

        # append info
        result.append(MetricInfo(ref, MetricKind(kind), MetricType(type_), name, size))

    return result


async def describe_metric(s: Session, ref: int, timeout: float = 5.0) -> MetricLayout:
    """Return the layout of the referenced metric."""

    # send command
    await s.send(_metrics_endpoint, pack("oo", 1, ref), 0)

    # prepare lists
    keys = {}
    values = {}

    while True:
        # receive reply
        reply, ack = await s.receive(_metrics_endpoint, True, timeout)
        if ack:
            break

        # verify reply
        if len(reply) < 1:
            raise RuntimeError("invalid reply")

        # handle key
        if reply[0] == 0:
            # verify reply
            if len(reply) < 3:
                raise RuntimeError("invalid reply")

            # add key
            keys[reply[1]] = reply[2:].decode()
            values[reply[1]] = {}

            continue

        # handle value
        if reply[0] == 1:
            # verify reply
            if len(reply) < 4:
                raise RuntimeError("invalid reply")

            # check key index
            if reply[1] not in values:
                raise RuntimeError(f"invalid key index: {reply[1]}")

            # add value
            values[reply[1]][reply[2]] = reply[3:].decode()

            continue

        raise RuntimeError("invalid reply")

    return MetricLayout(
        keys=[keys[num] for num in sorted(keys)],
        values=[
            [values[num][sub] for sub in sorted(values[num])] for num in sorted(keys)
        ],
    )


async def read_metrics(s: Session, ref: int, timeout: float = 5.0) -> bytes:
    """Return the raw values of the referenced metric."""

    # send command
    await s.send(_metrics_endpoint, pack("oo", 2, ref), 0)

    # receive reply
    reply, _ = await s.receive(_metrics_endpoint, False, timeout)

    return reply or b""


async def read_long_metrics(s: Session, ref: int, timeout: float = 5.0) -> List[int]:
    """Return the values of the referenced long metric."""

    return _convert(await read_metrics(s, ref, timeout), "i", 4)


async def read_float_metrics(s: Session, ref: int, timeout: float = 5.0) -> List[float]:
    """Return the values of the referenced float metric."""

    return _convert(await read_metrics(s, ref, timeout), "f", 4)


async def read_double_metrics(
    s: Session, ref: int, timeout: float = 5.0
) -> List[float]:
    """Return the values of the referenced double metric."""

    return _convert(await read_metrics(s, ref, timeout), "d", 8)


def _convert(data: bytes, code: str, width: int) -> list:
    # verify length
    if len(data) % width != 0:
        raise RuntimeError(f"invalid metric payload length: {len(data)}")

    return list(struct.unpack(f"<{len(data) // width}{code}", data))
